### PR TITLE
Add basic back-pressure to CockpitPipe and CockpitStream

### PR DESCRIPTION
**Pull request by stefwalter**
I'm landing this seperately to make finding bugs and reviewing easier. On its own this really does nothing, but it changes some core pieces of IO that get used frequently ... and we need real assistance from the bots to validate that this change in and of itself does not introduce regressions.

This adds back pressure and throttling to CockpitPipe and CockpitStream. They have the ability to trigger a "pressure" signal when their output queue is too full, and also consume a "pressure" signal to control their input flow.

### CODE REVIEW

**Comment by stefwalter** [comment]
Investigating check-memory failures.

**Comment by stefwalter** [comment]
Fixed timing problems in the tests.

**Comment by stefwalter** [comment]
The ```cockpit-bridge``` in ```test-channelresponse``` is getting killed with signal 9 for no easily apparent reason. Investigating that.

File: src/bridge/cockpitinteracttransport.c [hunk 1]
```diff
@@ -102,6 +102,9 @@ on_pipe_read (CockpitPipe *pipe,
         }
       g_bytes_unref (message);
     }
+
+  if (end_of_data)
+    cockpit_pipe_close (self->pipe, NULL);
```
Comment (allisonkarlitskaya) [inline src/bridge/cockpitinteracttransport.c]: unrelated change? (plus 2 more similar ones below)

File: src/common/cockpitpipe.c [hunk 2]
```diff
@@ -925,14 +1010,25 @@ _cockpit_pipe_write (CockpitPipe *self,
...
       return;
     }
 
+  before = self->priv->out_queued;
+  g_return_if_fail (G_MAXSIZE - size > self->priv->out_queued);
+  self->priv->out_queued += size;
   g_queue_push_tail (self->priv->out_queue, g_bytes_ref (data));
 
+  /*
+   * If we have two much data queued, and are controlling another flow
```
Comment (allisonkarlitskaya) [inline src/common/cockpitpipe.c]: "two much" :)

File: src/common/cockpitpipe.c [hunk 3]
```diff
@@ -925,14 +1010,25 @@ _cockpit_pipe_write (CockpitPipe *self,
...
+   * tell it to stop sending data, each time we cross over the high bound.
+   */
+  if (before < QUEUE_PRESSURE && self->priv->out_queued >= QUEUE_PRESSURE)
```
Comment (allisonkarlitskaya) [inline src/common/cockpitpipe.c]: What do you think would happen if we immediately applied backpressure as soon as we had even 1 item in the queue, and didn't realise it until the queue was empty again?

ie: QUEUE_PRESSURE = 1

File: src/common/cockpitpipe.c [hunk 4]
```diff
@@ -1554,3 +1652,58 @@ cockpit_pipe_get_environ (const gchar **input,
...
+  g_return_if_fail (COCKPIT_IS_PIPE (self));
+
+  if (self->priv->pressure)
+    {
+      g_signal_handler_disconnect (self->priv->pressure, self->priv->pressure_sig);
+      g_object_remove_weak_pointer (G_OBJECT (self->priv->pressure), (gpointer *)&self->priv->pressure);
+      self->priv->pressure = NULL;
+    }
+
+  self->priv->pressure_sig = 0;
```
Comment (allisonkarlitskaya) [inline src/common/cockpitpipe.c]: dubious.  In case the object is removed via weak pointer notification, pressure_sig will contain junk.  Your code seems safe against that possibility (ie: it always only checks ->pressure != NULL) so why do this unnecessary and confusing cleanup here?

File: src/common/cockpitpipetransport.c [hunk 5]
```diff
@@ -86,6 +86,9 @@ on_pipe_read (CockpitPipe *pipe,
   CockpitPipeTransport *self = COCKPIT_PIPE_TRANSPORT (user_data);
   cockpit_transport_read_from_pipe (COCKPIT_TRANSPORT (self), self->name,
                                     pipe, &self->closed, input, end_of_data);
+
+  if (end_of_data)
+    cockpit_pipe_close (self->pipe, NULL);
```
Comment (allisonkarlitskaya) [inline src/common/cockpitpipetransport.c]: also not sure what this is for

File: src/common/test-pipe.c [hunk 6]
```diff
@@ -319,6 +330,107 @@ test_pid (TestCase *tc,
...
+
+  /*
+   * Now the queue is getting drained. At some point, it will be
+   * signaled that back pressure has been turned off
+   */
+  while (throttle == -1)
+    g_main_context_iteration (NULL, TRUE);
+
+  g_assert_cmpint (throttle, ==, 0);
+  g_assert_cmpint (echo_pipe->received->len, >, 10 * 1000);
```
Comment (allisonkarlitskaya) [inline src/common/test-pipe.c]: why 10*1000?

File: src/bridge/cockpitstream.c [hunk 7]
```diff
@@ -745,6 +790,19 @@ cockpit_stream_class_init (CockpitStreamClass *klass)
...
 
+  /**
+   * CockpitStream::pressure:
+   * @throttle: Pressure on or off
+   *
+   * Emitted when the websocket wants to give back-pressure to other feeding
+   * streams. It does this when its output queue is too long and should
+   * slow down.
+   */
+  cockpit_stream_sig_pressure = g_signal_new ("pressure", COCKPIT_TYPE_STREAM, G_SIGNAL_RUN_FIRST,
```
Comment (allisonkarlitskaya) [inline src/bridge/cockpitstream.c]: I don't like the adhoc way you're defining this same thing in multiple places and then effectively relying on ducktyping when you go to hook stuff up.  Why not use an interface here?

Similarly: you could have a second interface for the other side of it (ie: objects that can receive pressure signals) and use the interface code to avoid some duplication of the weak pointer and signal hookup code.

Comment (stefwalter) [reply src/bridge/cockpitstream.c]: >  Why not use an interface here?

Sure, will do. An interface can reduce code. But almost any use of signals in a GLib program is duck-typing. The WebSocket code has also been logically separate, and someone can combine that separately. So that part will remain duck typed ... but that does not affect this pull request.

> Similarly: you could have a second interface for the other side of it (ie: objects that can receive pressure signals) and use the interface code to avoid some duplication of the weak pointer and signal hookup code.

That's not an interface. That would be a multi-inheritance base class ... or at best a strange helper function that takes a partial struct (ie: a mixin).

**Comment by stefwalter** [comment]
Since the switch to interfaces just adds boilerplate code, with no type safety advantage, but a slight advantage of removing similar lines of  code ... I'll do it in a separate branch and pull request. And then we can compare the difference.

**Comment by stefwalter** [comment]
I've implemented this as an interface CockpitFlow, which both CockpitPipe and CockpitStream implement. This provides no real advantage here ... but in the future we could move more common signals and vfuncs into CockpitFlow ... so I'm okay with this.

All other review comments addressed.

I will squash once ready to merge.

**Comment by stefwalter** [comment]
@martinpitt Using interfaces here means there must be another massive pull request moving the ```src/websocket``` code into ```src/common```, so that the WebSocket code changes in #9596 and #9553 can use CockpitFlow instead of the "duck-typing" that makes @allisonkarlitskaya queasy.

A bunch of churn due to 'sed'. Are you okay with that? If not, we should stick to the non-interface mechanism and drop the FIXUP commits here.

**Comment by martinpitt** [comment]
Stef Walter [2018-07-06 19:02 +0000]:
> @martinpitt Using interfaces here means there must be another
> massive pull request moving the ```src/websocket``` code into
> ```src/common```, so that the WebSocket code changes in #9596 and
> #9553 can use CockpitFlow instead of the "duck-typing" that makes
> @allisonkarlitskaya queasy.

While the size of the patch might be large, it should be rather simple
to understand, as it's just a type/declaration change. I'm fine with
that, so if Lis and you agree that a proper interface is cleaner,
please go ahead. (It does sound more robust).

**Comment by stefwalter** [comment]
> While the size of the patch might be large, it should be rather simple to understand, as it's just a type/declaration change. I'm fine with that, so if Lis and you agree that a proper interface is cleaner, please go ahead. (It does sound more robust).

We had an artificial barrier between the WebSocket code and the rest of the code (both by its location in the tree, its dependencies, and it's naming). Rather than move the WebSocket code, I've just started to break down that barrier. That should be simpler.

File: src/common/test-pipe.c (hunk 6 above)
Comment (allisonkarlitskaya) [reply src/common/test-pipe.c]: I guess I wanted to say "are you sure that this can never be 0, and if you're sure, then shouldn't you also be sure that it would be more than one block?"

File: src/common/cockpitflow.c [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/**
+ * CockpitFlow:
+ *
+ * An interface representing a bidirectional flow. Implementors are
```
[the remark this answers is not in the document]
Comment (stefwalter) [reply src/common/cockpitflow.c]: Understood. Nice summary of the discussion on IRC.

File: src/common/cockpitflow.c [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+cockpit_flow_throttle (CockpitFlow *flow,
+                       CockpitFlow *controlling)
+{
+  CockpitFlowIface *iface;
+
+  g_return_if_fail (COCKPIT_IS_FLOW (flow));
+  g_return_if_fail (controlling == NULL || COCKPIT_IS_FLOW (controlling));
+
+  iface = COCKPIT_FLOW_GET_IFACE (flow);
+  g_assert (iface->throttle != NULL);
```
[the remark this answers is not in the document]
Comment (stefwalter) [reply src/common/cockpitflow.c]: As suggested, I have used g_return_if_fail ... which is a g_critical error when reached.

File: src/bridge/cockpitstream.c [hunk 10]
```diff
@@ -30,6 +30,13 @@
  * CockpitStream:
  *
  * A stream with queued input and output based on top of a GIOStream
+ *
+ * This stream can do flow control in two ways:
+ *
+ *  - Its input can be throttled, it can listen to a "pressure" signal
+ *    from another object passed into cockpit_stream_throttle()
+ *  - It can optionally control another flow, by emitting a "pressure" signal
+ *    when it's output queue is too large
```
[the remark this answers is not in the document]
Comment (stefwalter) [reply src/bridge/cockpitstream.c]: Fixed both cases.

**Comment by stefwalter** [comment]
Made changes for all review points.

**Comment by stefwalter** [comment]
Rebased.

File: src/common/cockpitflow.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (stefwalter) [reply src/common/cockpitflow.c]: FWIW, only one implementation of this interface (across 8 or 9) is uni directional.

File: src/common/cockpitpipe.h [hunk 11]
```diff
@@ -57,6 +57,9 @@ struct _CockpitPipeClass {
 
   void        (* close)       (CockpitPipe *pipe,
                                const gchar *problem);
+
+  void        (* pressure)    (CockpitPipe *pipe,
```
Comment (allisonkarlitskaya) [inline src/common/cockpitpipe.h]: did you forget to drop this?

File: src/bridge/cockpitstream.h [hunk 12]
```diff
@@ -56,6 +56,9 @@ struct _CockpitStreamClass {
 
   void        (* close)       (CockpitStream *pipe,
                                const gchar *problem);
+
+  void        (* pressure)    (CockpitStream *stream,
```
Comment (allisonkarlitskaya) [inline src/bridge/cockpitstream.h]: ditto.
